ruby26: Fix build issues on 10.4, 10.5.

fhgwright opened the pull request:
This includes a patch for some OS version issues common to earlier
Ruby versions:

The OS version conditional in dln.c is comparing an undefined macro to
a sometimes undefined (and essentially useless) macro.  However, the
relevant code doesn't seem to be built for any OS version, making the
bug irrelevant.  Nevertheless, it's now fixed, since another, similar
fix was needed, anyway, and it might become relevant if something else
changes.

The OS version conditionals in error.c were using another useless
macro for the comparison, this time with a real effect, though only in
a couple of messages related to crash reports on < 10.6.  This is now
fixed.

It also includes an updated patch for a 10.4 issue:

The context register naming issue on 10.4 was sort of fixed upstream
but tied to an inappropriate config flag.  The patched version uses an
SDK version macro to make the decision.

A newly discovered issue, common to some older Ruby versions, is:

The 10.5 x86_64 build fails if libunwind-headers is active.  Since
this is an unusual configuration, and the need for libunwind-headers
seems to be waning, it's not worth fixing "right".  This just adds the
usual failure for such cases.  It's unknown why this only affects 10.5
x86_64 (and perhaps 10.4 x86_64, though that's untested).

The remaining issues were new in ruby26:

The syntax-checking compiler invocations in transform_mjit_header.rb
require a C99 or later compiler, but don't include compiler flags from
the build.  There's code to detect the non-C99 case and add -std=c99
to the real uses, but it relies on matching the error message from the
test case.  Recent versions of gcc use locale-dependent quotes which
may break such comparisons, depending on the current locale settings.
The fix recommended by gcc documentation is to run it with LC_ALL=C to
get traditional ASCII quotes.  This adds code to apply that
environment to the compiler test invocation.

For some strange reason, using gcc-4.2 for Ruby's coroutine-related
assembler code on x86_64 passes "-arch i386" to the assembler, thereby
causing failures on x86_64 register references.  To fix that, cases
with gcc-4.2 and x86_64 targeted now pass the correct architecture in
ASFLAGS.

Also corrects a small typo in the description.

Also fixes some lint errors in the Portfile.

Since this affects the installed code for < 10.6, it includes a
revbump.

Universal builds still fail on 10.4 i386 due to a weird configure bug,
and 10.4 ppc has broken universal dependencies.

TESTED:
Built -universal on 10.4-10.5 ppc, 10.4-10.6 i386, and 10.6-10.15
x86_64.  Built +universal on 10.5-10.13.

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
Mac OS X 10.4.11 8S165, PPC, Xcode 2.5 8M2558
Mac OS X 10.4.11 8S2167, i386, Xcode 2.5 8M2558
Mac OS X 10.5.8 9L31a, PPC, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, i386, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, x86_64, Xcode 3.1.4 9M2809
Mac OS X 10.6.8 10K549, i386, Xcode 3.2.6 10M2518
Mac OS X 10.6.8 10K549, x86_64, Xcode 3.2.6 10M2518
Mac OS X 10.7.5 11G63, x86_64, Xcode 4.6.3 4H1503
OS X 10.8.5 12F2560, x86_64, Xcode 5.1.1 5B1008
OS X 10.9.5 13F1911, x86_64, Xcode 6.2 6C131e
OS X 10.10.5 14F2511, x86_64, Xcode 7.2 7C68
OS X 10.11.6 15G22010, x86_64, Xcode 8.1 8B62
macOS 10.12.6 16G2136, x86_64, Xcode 9.2 9C40b
macOS 10.13.6 17G14042, x86_64, Xcode 10.1 10B61
macOS 10.14.6 18G9323, x86_64, Xcode 11.3.1 11C505
macOS 10.15.7 19H2, x86_64, Xcode 12.4 12D4e

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [N/A] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [N/A] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

kencu commented:
I use something like this:

https://github.com/kencu/TigerPorts/blob/master/lang/ruby25/files/patch-ruby25-tigeri386.diff

This PR looks very complicated.

kencu commented:
BTW, I locally overrode the changes made last year in the rubys for Tiger, as I think the patch I referenced is the more correct fix for those Tiger hiccups.

fhgwright commented:
@kencu:
> I use something like this:
> 
> https://github.com/kencu/TigerPorts/blob/master/lang/ruby25/files/patch-ruby25-tigeri386.diff
> 
> This PR looks very complicated.

Why am I not surprised that you chimed in on this ? :-)

As far as this particular fix goes, mine is simpler than yours.  The overall PR has a lot because there were multiple issues to fix, all of which are explained in the commit message.  I originally had multiple commits (starting with one inheriting the patches from `ruby22-25`), but squashed them for the PR in keeping with MacPorts policy.

My Tiger patch simply extends the scope of the existing conditional for an optional feature, to cover a non-working case that it had missed.  Because of the nature of the change, a successful build is a sufficient test.  Your fix involves actually making the context stuff work on 10.4 i386, which then requires actually testing that feature.  I don't know how to test that, and of course there's no "port test ruby26" at all.  Do you have a test for it?  _I_ don't want to push a change that _I_ don't know how to test properly, and it doesn't make sense to make the perfect the enemy of the good by leaving `ruby26` in its current state where it (and consequently asciidoctor) is broken in all 10.4/10.5 cases.  You can always follow up with a better fix, if you have a way to properly test it, though if it's really that important, it ought to be fixed for PPC as well, particularly given that 10.4 ppc is probably far more common than 10.4 i386.

About a year ago, I attacked `ruby24` because it was needed for `asciidoctor`.  I figured that as long as I was working on `ruby24`, I might as well fix as many other ruby versions as possible.  I wound up fixing `ruby22-25`, but `ruby26-27` presented additional problems that looked hard to fix, so I left it at that.  But then someone moved `asciidoctor` to `ruby26` for arm64 compatibility, reintroducing the old 10.4/10.5 breakage.  I was right about "hard to fix", since the `transform_mjit_header.rb` bug was a royal PITA to track down, including a time when the bug was literally staring me in the face and I didn't see it.  After the dust has settled, I may revisit `ruby27` in light of what I've learned.


BTW, it occurs to me that the issue with `libunwind-headers` on 10.5 x86_64 may also apply to 10.4 x86_64, but I don't have a VM set up to test that case.  I believe that 10.4 x86_64 is theoretically possible, but too weird to be worth the hassle.  As it is, both 10.5 x86_64 and 10.6 i386 have proven to be "interesting".

fhgwright commented:
I rebased this with a new revbump, due to the collision with the libffi-induced revbump.

kencu commented:
I've been running ruby26 on Leopard and Tiger for several years, using 2 simple fixes in this repo:

https://github.com/kencu/TigerPorts/tree/master/lang/ruby26

please take a look. It's much much much much simpler than this, and I think a better fix.

kencu commented:
converted to draft because I don't think this should be accidentally merged until other options are looked at that are much simpler to maintain going forward

kencu commented:
OK, so according my last test, ruby26 is already working on 10.4 PPC, 10.5 PPC, and (I think) 10.5 Intel (have to triple check that) and to fix the build on 10.4 Intel I did only this:

```
    if {${os.major} < 10} {
        depends_build-append    port:gmake
        build.cmd               ${prefix}/bin/gmake
        configure.args-append   --disable-dtrace
        configure.cflags-append -std=c99
    }

    if {${os.major} == 8} {
        configure.cppflags-append -DCPU_SUBTYPE_MASK=0xff000000

+        # force old register names #11509 
+       configure.cppflags-append -D__DARWIN_UNIX03=0
    }
```
I didn't push that last tiny bit to MacPorts as yet.

None of the rest of the PR was needed, and the test suite passes nicely for me

So I will have to see what is up here with this PR and if something has changed somehow to require all this.

kencu commented:
> @kencu:
> 
> > I use something like this:
> > https://github.com/kencu/TigerPorts/blob/master/lang/ruby25/files/patch-ruby25-tigeri386.diff
> > This PR looks very complicated.
> 
> Why am I not surprised that you chimed in on this ? :-)

Nice.

I simply don't think you have so far understood the different register names in Tiger i386 vs newer systems (a very simple change) that is much much simpler than your ruby fixes to date.

kencu commented:
actually I change my mind. I don't use any of these ruby ports in MacPorts on older systems anyway, I use my own which work perfectly and I can understand them, so it matters not at all to me if this gets merged or not as I won't be using it anyway.

Fire away.

kencu commented:
https://github.com/macports/macports-ports/pull/12181

kencu commented:
Fred, I was using a slightly older ruby26. I'll try my fix on the current version and see if it still works now.

kencu commented:
everywhere else in MacPorts, we use the proper regnames fix for the Tiger SDK.

eg.
https://github.com/macports/macports-ports/commit/f9641f5f5cb7287e6787b566945fff460a80bc39#diff-c5cf17e504a1536e91626cc4225a8486d18080f47a3e3f6f1653cb4fddadea37

can you please take a look at that, and other similar fixes throughout MacPorts, eg cctools, and possibly chang yours to match that, rather than incorrectly blicking out the entire section with a system version block?

I would be most appreciative if you did thst here, and whenever else you come across this issue, and it will prevent me from having to come back later and fix it, or worse judt ignore it and leave it wrong.

kencu commented:
typos due to ipad fingers :) I'm splurging for a new ipad this Xmas, and then I hopefully will be able to edit in github, which I cannot presently do.

fhgwright commented:
I'll take a look, but probably not today.

Overall, it isn't just about the register names - every change here represents something I found, fixed, and explained.

Calling the `libunwind-headers` issue new in `ruby26` is actually inaccurate - it's just that I hadn't encountered it before (it also affects `ruby[2-5]` at least).  Whether I currently have `libunwind-headers` active has tended to depend on whether I last built `cctools` or some version of `gcc`, though now I'm starting to make a point of having it active when testing ports in order to catch such things.  For `ruby2[2-6]` (and perhaps others), I have no idea why it only matters for 10.5 x86_64, but that's what I see.

fhgwright commented:
I've updated this.

The 10.4 context register bug had an upstream fix (that happened later than when I was working on this last year), but it was tied to `__DARWIN_UNIX03`, which is normally set up by `cdefs.h`, but not in an appropriate way for this use.  Defining it directly has potentially far-reaching side effects, so I replaced that check with an SDK version check.

I removed the `configure.ac` consistency patch, since rerunning `autoconf` results in a completely broken `configure` anyway.  I added a comment to that effect in the Portfile.  Note that the current `configure` patch isn't upstreamable, since patching the output of `autoconf` isn't acceptable as an upstream fix.

It turns out that `configure` isn't the only generated file that doesn't rebuild correctly on the Mac, and make rebuilds them (often incorrectly) any time they look outdated based on mtime comparisons.  Fortunately, the tarball has consistent mtimes that don't suffer from this, but it causes intermittent failures when working directly in a git repo, due to git's propensity for clobbering mtimes (which will never be fixed since Linus considers it to be a feature).

There are at least two bugs in the tests that make it premature to enable the test phase:

1. Something which isn't even strictly a test issue is that using `Dir.mktmpdir` to create Unix domain socket paths easily exceeds the hard-coded OS socket-name length limit when the current directory has a verbose pathname (as is usually true of MacPorts builds).

2. I've never seen the tests run to completion without the test runner hanging.

@kencu 

> everywhere else in MacPorts, we use the proper regnames fix for the Tiger SDK.
> 
> eg.
> [f9641f5#diff-c5cf17e504a1536e91626cc4225a8486d18080f47a3e3f6f1653cb4fddadea37](https://github.com/macports/macports-ports/commit/f9641f5f5cb7287e6787b566945fff460a80bc39#diff-c5cf17e504a1536e91626cc4225a8486d18080f47a3e3f6f1653cb4fddadea37)

Note that that patch is not in an upstreamable form, since it relies on a conditionally applied patch.

kencu commented:
> Note that that patch is not in an upstreamable form, since it relies on a conditionally applied patch.

Tiger is unusable without MacPorts' newer compilers, cctools, ld64, meson, and gcc7/libgcc7 integration anyway, so I am not going to hold people's feet to the fire about Tiger patches being upstreamable -- I'd be too embarrassed to open the PR :>

kencu commented:
this looks good now, thanks for redoing it.

@kimuraw -- LGTM.

kencu commented:
Thanks, @fhgwright !
